### PR TITLE
Role-Specialized Surface Heads: dedicated SRF per foil role

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Replace the single shared SRF with **three role-specialized surface refinement heads**, each optimized for a specific prediction regime:

1. **Single-foil head** — used for all nodes in single-foil samples
2. **Tandem-fore head** — used for fore-foil nodes in tandem samples
3. **Tandem-aft head** — used for aft-foil nodes in tandem samples

**Why this helps p_tan:** The current shared SRF must simultaneously learn three fundamentally different pressure prediction tasks: (a) single-foil Cp (relatively easy, well-represented), (b) tandem fore-foil Cp (moderate — modified by upstream aft-foil influence), and (c) tandem aft-foil Cp (hardest — dominated by wake interference, separation shifts, and downwash). A shared MLP is forced to compromise between these regimes. Role specialization lets the aft-foil head dedicate ALL its capacity to learning the interference patterns that drive p_tan, without being diluted by the easier single-foil patterns.

**Key insight from prior experiments:** Surface cross-attention (#2372) showed genuine tandem benefit (p_tan -3.7%) but hurt single-foil (p_in +9.2%). The fix was tandem-conditional application. This experiment takes that principle further: not just conditional APPLICATION but conditional ARCHITECTURE — each regime gets its own optimized head.

### Architecture

Replace `SurfaceRefinementHead` with a `RoleSpecializedSRF` module containing three independent SRF heads:

```python
class RoleSpecializedSRF(nn.Module):
    def __init__(self, hidden_dim, output_dim, n_layers=3):
        super().__init__()
        # Three independent SRF heads (same architecture as current SRF)
        self.single_head = SurfaceRefinementHead(hidden_dim, output_dim, n_layers)
        self.tandem_fore_head = SurfaceRefinementHead(hidden_dim, output_dim, n_layers)
        self.tandem_aft_head = SurfaceRefinementHead(hidden_dim, output_dim, n_layers)
    
    def forward(self, x, foil_role, is_tandem):
        # foil_role: 0=single, 1=tandem-fore, 2=tandem-aft
        # Route nodes to the appropriate head
        out = torch.zeros_like(x[..., :output_dim])
        
        single_mask = (foil_role == 0)
        fore_mask = (foil_role == 1)
        aft_mask = (foil_role == 2)
        
        if single_mask.any():
            out[single_mask] = self.single_head(x[single_mask])
        if fore_mask.any():
            out[fore_mask] = self.tandem_fore_head(x[fore_mask])
        if aft_mask.any():
            out[aft_mask] = self.tandem_aft_head(x[aft_mask])
        
        return out
```

### Implementation Details

Add flag: `--role_specialized_srf`

When `--role_specialized_srf` is active:
- Replace the single SRF with `RoleSpecializedSRF` containing 3 independent SRF heads
- Each head has the SAME architecture as the current SRF (3 layers, same hidden dim) — total params ~3× current SRF
- Route nodes based on sample type and foil role:
  - Single-foil samples: ALL nodes → `single_head`
  - Tandem samples, fore-foil nodes → `tandem_fore_head`
  - Tandem samples, aft-foil nodes → `tandem_aft_head`
- The `foil_role` information should already be available in the batch data (check `foil_id` or similar field). If not, derive it from the `is_tandem` flag and node positions (fore-foil nodes have lower x-coordinates)
- The backbone (all TransolverBlocks) remains SHARED — only the SRF output is specialized
- Loss computation: each head's output contributes to the loss for its respective nodes. No change to the loss formulation

### Optimization Details

- Use the same lr (2e-4) for all three heads initially
- Each head sees fewer training samples than the current shared SRF:
  - `single_head`: ~50% of samples (all single-foil)
  - `tandem_fore_head`: ~50% × 50% = ~25% of total nodes
  - `tandem_aft_head`: ~50% × 50% = ~25% of total nodes
- The aft head sees the LEAST data but has the HARDEST task. If p_tan doesn't improve, try 2× lr for the aft head in v2

### Parameter Count

Current SRF: ~110K params (3 layers, 192 hidden)
Role-Specialized: ~330K params (3 × 110K)
Total model: ~3.5M → ~3.7M (minimal increase relative to backbone)
Epoch time: should be IDENTICAL — the masked routing avoids redundant computation

### Training Setup

```bash
cd cfd_tandemfoil && python train.py \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only \
  --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity \
  --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --role_specialized_srf \
  --wandb_group role-specialized-srf \
  --seed 42
```

Run 2 seeds (42, 73) as standard.

### What to Watch

- **p_tan** is the PRIMARY metric — the aft-foil specialist should help
- **p_in** should stay similar or improve (the single-foil head is optimized for single-foil)
- Compare per-foil metrics if available: is the aft-foil head actually learning better than the shared SRF?
- Check that the routing is correct: verify each head only processes its designated nodes (log node counts per head)
- Epoch time should NOT increase significantly (masked routing, not parallel forward passes)

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.872** | < 11.872 |
| p_oodc | 7.459 | < 7.459 |
| **p_tan** | **26.319** | < 26.319 |
| **p_re** | **6.229** | < 6.229 |

Baseline PR: #2357 (Vortex-Panel Induced Velocity)

Reproduce baseline:
```
cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Role-Specialized Surface Refinement Heads

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Delta |
|--------|---------|---------|------------|----------|-------|
| **p_in** | 12.4 | 12.3 | **12.35** | 11.872 | **+4.0%** |
| p_oodc | 7.6 | 7.2 | **7.40** | 7.459 | **-0.8%** |
| **p_tan** | 27.5 | 28.6 | **28.05** | 26.319 | **+6.6%** |
| p_re | 6.2 | 6.3 | **6.25** | 6.229 | **+0.3%** |

### Training Details

- **Command:**
```bash
cd cfd_tandemfoil && python train.py \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only \
  --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity \
  --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --role_specialized_srf \
  --wandb_group role-specialized-srf \
  --seed {42,73}
```
- **Peak VRAM:** 40.4–40.7 GB (well within 96 GB budget)
- **Epoch time:** ~70–71s (identical to baseline — masked routing avoids redundant computation as expected)
- **Epochs completed:** 151–152 (wall-clock limit at 180 min)
- **Role-specialized SRF params:** 127,497 (3 × 42,499 — 3× the standard SRF)
- **W&B runs:** [seed42/jlo9yryu](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/jlo9yryu) | [seed73/qxnrmurv](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/qxnrmurv)

### What Happened

**Negative result.** Role-specialized SRF did not improve any primary metric and regressed on the two most important ones:

- **p_in +4.0%**: The single-foil head saw only ~50% of surface nodes compared to the shared SRF. Despite having dedicated capacity for single-foil patterns, the reduced data volume hurt generalization.
- **p_tan +6.6%**: This is the opposite of what we hoped. The tandem-aft head, which should have been the main beneficiary, saw only ~25% of total surface data — the hardest task with the least data. The shared SRF's cross-regime transfer actually helped p_tan more than specialization.
- **p_oodc -0.8%** and **p_re +0.3%**: Approximately flat — these are within noise.

**Key insight:** The shared SRF's ability to transfer learned features across foil roles is actually beneficial. Splitting into specialized heads fragments the data each head sees without providing enough capacity benefit to compensate. The current SRF architecture (128-hidden, 2 layers) is small enough that "regime compromise" isn't the bottleneck — data volume is.

### Suggested Follow-ups

1. **Shared backbone + role-specific last layer:** Instead of fully independent heads, keep a shared SRF backbone (first 1-2 layers) with role-specific output projections. This preserves feature transfer while allowing output specialization.
2. **Soft routing / mixture-of-experts approach:** Use learned soft gating instead of hard role masks. Let the model learn how to blend predictions from multiple heads, potentially with the shared head as a strong prior.
3. **Aft-foil-only specialization with higher lr:** The prior result from #2372 (cross-attention p_tan -3.7%) suggests that tandem-specific processing helps, but only when applied additively on top of the shared prediction — not as a replacement for it.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2381 because ADVISOR: Clear dead end — p_in +4.0%, p_tan +6.6%. Good analysis on why: fragmenting SRF data reduces each head's volume without compensating capacity benefit. The shared SRF's cross-regime transfer is genuinely helpful. Key insight saved for future reference: specialization via splitting hurts; specialization via adding (cross-attention, constraints) helps. Closing — assigning you a new experiment..
